core/src/main/java/org/elasticsearch/ingest/processor/DeduplicateProcessor.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.ingest.core.AbstractProcessor;
+import org.elasticsearch.ingest.core.AbstractProcessorFactory;
+import org.elasticsearch.ingest.core.IngestDocument;
+import org.elasticsearch.ingest.core.ConfigurationUtils;
+
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * Processor that removes duplicate entries from an array.
+ * Throws exception is the specified field is not an array.
+ */
+public final class DeduplicateProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "dedupe";
+    public static final ParseField FIELD = new ParseField("field");
+    public static final ParseField ORDERED = new ParseField("ordered");
+    public static final boolean DEFAULT_ORDERED = true;
+
+    private final String field;
+    private final boolean ordered;
+
+    DeduplicateProcessor(String tag, String field, boolean ordered) {
+        super(tag);
+        this.field = field;
+        this.ordered = ordered;
+    }
+
+    String getField() {
+        return field;
+    }
+
+    @Override
+    public void execute(IngestDocument document) {
+        List<?> list = document.getFieldValue(field, List.class);
+        if (list == null) {
+            throw new IllegalArgumentException("field [" + field + "] is null, cannot deduplicate.");
+        }
+
+        if (list.size() <= 1) {
+            return;
+        }
+
+        if (lastType != null && lastType.equals("sort")) {
```

Comment:
I was curious about how new ingest processors are implemented, so looked into this PR. Regarding requiring "sort": what if lastType is "sort" but it was sorting another field? Something that might be caught by pipeline validation perhaps.